### PR TITLE
Feat/enable oauth url change

### PR DESCRIPTION
**Pull request by komret**
Merging these changes and redeploying `@trezor/auth-server` to staging should allow testing the default (authorisation code) Google OAuth flow.

1. Fixes allowed origins in `@trezor/auth-server`
2. Adds the option to change auth-server URL in debug settings
3. Adds handling for unexpected response from auth server

@vdovhanych I noticed the calls to auth server fail when Tor is on (it requests an onion address). How can we fix that?

**QA:** Authorising with Google should behave the same as previously from the user's perspective, but access is not limited to 1 hour (access token will always refresh if invalid). There is a new setting in debug to change between production/staging/localhost for the authorization server - you can keep the default (production value); staging does not work with Tor enabled. Please check that the user stays authorized with Google for more than 1 hour on desktop. You can confirm it by changing a label and seeing a successful request to https://suite-auth.trezor.io/google-oauth-refresh in DevTools. These changes should not affect Dropbox, filesystem storage.

### CODE REVIEW

**Comment by vdovhanych** [comment]
We probably could get an onion address on that, but I would do it without that for now and discuss it if we even want that. News API has a similar service running on the web that is handled somehow, so when tor is enabled, it still works.

File: packages/suite/src/views/settings/debug/OAuthApi.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+`;
+
+export const OAuthApi = () => {
+    const { setDebugMode } = useActions({
+        setDebugMode: suiteActions.setDebugMode,
+    });
+    const { debug } = useSelector(state => ({
+        debug: state.suite.settings.debug,
+    }));
+    const { anchorRef, shouldHighlight } = useAnchor(SettingsAnchor.InvityApi);
```
Comment (tomasklim) [inline packages/suite/src/views/settings/debug/OAuthApi.tsx]:
```suggestion
    const { anchorRef, shouldHighlight } = useAnchor(SettingsAnchor.OAuthApi);
```

File: packages/suite/src/views/settings/debug/OAuthApi.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import GoogleClient from '@suite/services/google';
+import { ActionColumn, ActionSelect, SectionItem, TextColumn } from '@suite-components/Settings';
+import * as suiteActions from '@suite-actions/suiteActions';
+import { useSelector, useActions } from '@suite-hooks';
+import type { OAuthServerEnvironment } from '@suite-types/metadata';
+import { useAnchor } from '@suite-hooks/useAnchor';
+import { SettingsAnchor } from '@suite-constants/anchors';
+
+const StyledActionSelect = styled(ActionSelect)`
+    min-width: 256px;
```
Comment (tomasklim) [inline packages/suite/src/views/settings/debug/OAuthApi.tsx:13]: This should be applied only to desktop/web. It "breaks" on mobile. (You can fix it also for invityApi)

File: packages/suite/src/services/google.ts [hunk 3]
```diff
@@ -208,14 +217,28 @@ class Client {
...
 
                 const json = await res.json();
-                Client.accessToken = json.access_token;
-                Client.refreshToken = json.refresh_token;
+                if (!json?.access_token || !json?.refreshToken) {
+                    throw new Error('Could not retrieve the tokens.');
+                } else {
+                    Client.accessToken = json.access_token;
+                    Client.refreshToken = json.refresh_token;
+                }
```
Comment (tomasklim) [inline packages/suite/src/services/google.ts]:
```suggestion
                Client.accessToken = json.access_token;
                Client.refreshToken = json.refresh_token;
```

File: packages/suite/src/constants/suite/anchors.ts [hunk 4]
```diff
@@ -35,6 +35,7 @@ export const enum SettingsAnchor {
     GithubIssue = '@debug-settings/github-issue',
     WipeData = '@debug-settings/wipe-data',
     InvityApi = '@debug-settings/invity-api',
+    OauthApi = '@debug-settings/oauth-api',
```
Comment (tomasklim) [inline packages/suite/src/constants/suite/anchors.ts]:
```suggestion
    OAuthApi = '@debug-settings/oauth-api',
```

File: packages/suite/src/services/google.ts [hunk 5]
```diff
@@ -208,14 +217,28 @@ class Client {
...
             } catch {
-                Client.authServerAvailable = false;
+                await Client.forceImplicitFlow();
             }
         }
     }
 
+    // when auth server is running, but returns an unexpected response, fall back to implicit flow
+    // TODO: this does not work if browser blocks pop-up windows and it opens two tabs/windows, there could be a better solution
```
Comment (tomasklim) [inline packages/suite/src/services/google.ts:231]: Is it possible to react on pop-up close? User changes his mind and wants to use another provider instead. But other options are disabled.
Screenshot 2022-07-14 at 11 13 18
r

**Review by tomasklim — DISMISSED** [review]
Localhost Web - First try works, second one does not
https://user-images.githubusercontent.com/33235762/178950838-035cd015-ea51-40a4-8c5a-134b93d535bc.mov

Desktop
I have to do it twice.
https://user-images.githubusercontent.com/33235762/178951740-2e52c4f6-8872-4c03-b51e-064b96de24f8.mov

File: packages/suite/src/views/settings/debug/OAuthApi.tsx (hunk 2 above)
Comment (matejkriz) [reply packages/suite/src/views/settings/debug/OAuthApi.tsx:13]: it's a small thing, so feel free to fix it, but no need to fix mobile UI for debug menu...

File: packages/suite/src/views/settings/debug/OAuthApi.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    return (
+        <SectionItem
+            data-test="@settings/debug/oauth-api"
+            ref={anchorRef}
+            shouldHighlight={shouldHighlight}
+        >
+            <TextColumn
+                title="API server"
+                description="Set the server url for generating Google refresh token"
```
Comment (matejkriz) [inline packages/suite/src/views/settings/debug/OAuthApi.tsx]: Would be nice to mention "labeling" here
```suggestion
                title="Google Auth server"
                description="Set the authorization server for labeling in Google Drive"
```

File: packages/auth-server/src/index.ts [hunk 7]
```diff
@@ -8,10 +8,10 @@ app.use(express.json());
...
 const corsOptions: CorsOptions = {
     origin: [
-        'trezor.io', // production web
-        'sldev.cz', // staging web
-        'localhost', // development web + all desktop
-        'trezoriovpjcahpzkrewelclulmszwbqpzmzgub37gbcjlvluxtruqad.onion', // onion address for production web (Suite does not work here now)
+        'https://suite.trezor.io', // production web
+        /\.sldev\.cz$/, // staging web
+        'http://localhost:8000', // development web
+        'https://trezoriovpjcahpzkrewelclulmszwbqpzmzgub37gbcjlvluxtruqad.onion', // onion address for production web (Suite does not work here now)
```
Comment (matejkriz) [inline packages/auth-server/src/index.ts]: it's not HTTPS over onion
```suggestion
        'http://trezoriovpjcahpzkrewelclulmszwbqpzmzgub37gbcjlvluxtruqad.onion', // onion address for production web (Suite does not work here now)
```

File: packages/suite/src/views/settings/debug/OAuthApi.tsx (hunk 1 above)
Comment (komret) [reply packages/suite/src/views/settings/debug/OAuthApi.tsx]: Fixed in 469efbf93ea70aef646477c7b77bf76e1481be54

File: packages/suite/src/views/settings/debug/OAuthApi.tsx (hunk 2 above)
Comment (komret) [reply packages/suite/src/views/settings/debug/OAuthApi.tsx:13]: I also think this is OK for debug settings and 256px is not too long to break most mobile layouts.

File: packages/suite/src/services/google.ts (hunk 3 above)
Comment (komret) [reply packages/suite/src/services/google.ts]: Done in 469efbf93ea70aef646477c7b77bf76e1481be54

File: packages/suite/src/constants/suite/anchors.ts (hunk 4 above)
Comment (komret) [reply packages/suite/src/constants/suite/anchors.ts]: Done in 469efbf93ea70aef646477c7b77bf76e1481be54

File: packages/suite/src/views/settings/debug/OAuthApi.tsx (hunk 6 above)
Comment (komret) [reply packages/suite/src/views/settings/debug/OAuthApi.tsx]: Done in 469efbf93ea70aef646477c7b77bf76e1481be54

File: packages/auth-server/src/index.ts (hunk 7 above)
Comment (komret) [reply packages/auth-server/src/index.ts]: Done in 69a2b8584072e289bf020a110e1093185587af24

File: packages/suite/src/services/google.ts (hunk 5 above)
Comment (komret) [reply packages/suite/src/services/google.ts:231]: I noticed it stops loading with a small delay after to pop-up is closed manually. It only works on web, though, and could be a bit faster. I will look into it, but I would leave it outside of this PR.

**Comment by komret** [comment]
@tomasklim The auth-server does not fully work yet, authorisation should be smooth once it does. The double login experienced on desktop is due to the server failing authentication and the app falling back to implicit flow (thus asking permission again). This should not happen in practice, though.

**Review by matejkriz — APPROVED** [review]

File: packages/suite/src/views/settings/debug/OAuthApi.tsx (hunk 2 above)
Comment (tomasklim) [reply packages/suite/src/views/settings/debug/OAuthApi.tsx:13]: Ok nevermind, I will fix it in one of my future small things PRs

**Pull request merged**